core/io/file_access_pack.cpp

```diff
@@ -248,14 +181,150 @@ bool PackedSourcePCK::try_open_pack(const String &p_path, bool p_replace_files,
+	uint32_t magic;
+	if (pck_off != 0) {
+		// Search for the header, in case PCK start and section have different alignment.
+		for (int i = 0; i < 8; i++) {
+			f->seek(pck_off);
+			magic = f->get_32();
+			if (magic == PACK_HEADER_MAGIC) {
+#ifdef DEBUG_ENABLED
+				print_verbose("PCK header found in executable pck section, loading from offset 0x" + String::num_int64(pck_off - 4, 16));
+#endif
+				pck_header_found = true;
+				break;
+			}
+			pck_off++;
+		}
+	}
+
+	// Search for the header at the end of file - self contained executable.
+	if (!pck_header_found) {
+		f->seek_end();
+		f->seek(f->get_position() - 4);
+		magic = f->get_32();
+
+		if (magic == PACK_HEADER_MAGIC) {
+			f->seek(f->get_position() - 12);
+			uint64_t ds = f->get_64();
+			f->seek(f->get_position() - ds - 8);
+			magic = f->get_32();
+			if (magic == PACK_HEADER_MAGIC) {
+#ifdef DEBUG_ENABLED
+				print_verbose("PCK header found at the end of executable, loading from offset 0x" + String::num_int64(f->get_position() - 4, 16));
+#endif
+				pck_header_found = true;
+			}
+		}
+	}
+
+	if (!pck_header_found) {
+		return false;
+	}
+
+	return process_pck(p_path, f, 0);
+}
+
+bool PCKFile::try_open_pack(const String &p_path, uint64_t p_offset) {
+	if (is_open()) {
+		close();
+	}
+
+	Ref<FileAccess> f = FileAccess::open(p_path, FileAccess::READ);
+	if (f.is_null()) {
+		return false;
+	}
+
+	// Search for the header at the start offset - standalone PCK file.
+	f->seek(p_offset);
+	uint32_t magic = f->get_32();
+	if (magic != PACK_HEADER_MAGIC) {
+		return false;
+	}
+
+	return process_pck(p_path, f, p_offset);
+}
+
+bool PCKFile::is_open() {
+	if (file.is_valid()) {
+		return file->is_open();
+	} else {
+		return false;
+	}
+}
+
+void PCKFile::close() {
+	files.clear();
+	file = Ref<FileAccess>();
+}
+
+Vector<Pair<String, PackedData::PackedFile>> PCKFile::get_files() {
+	Vector<Pair<String, PackedData::PackedFile>> result;
+	for (auto [k, v] : files) {
+		result.append({ k, v });
 	}
+	return result;
+}
 
+Ref<FileAccess> PCKFile::get_file(const String &p_path, bool p_case_sensitive) {
+	if (!p_case_sensitive) {
+		WARN_PRINT("foo");
+		String filepath = case_insensitive_filenames[p_path.to_lower()];
+		return memnew(FileAccessPack(filepath, files[filepath]));
+	}
+
+	return memnew(FileAccessPack(p_path, files[p_path]));
+}
+
+bool PCKFile::file_exists(const String &p_path, bool p_case_sensitive) {
+	return p_case_sensitive ? files.has(p_path) : case_insensitive_filenames.has(p_path.to_lower());
+}
+
+//////////////////////////////////////////////////////////////////
+
+bool PackedSourcePCK::try_open_pack(const String &p_path, bool p_replace_files, uint64_t p_offset) {
+	if (!file.try_open_pack(p_path, p_offset)) {
+		if (!file.try_open_embedded(p_path)) {
+			return false;
+		}
+	}
+
+	for (auto [path, f] : file.get_files()) {
+		PackedData::get_singleton()->add_path(p_path, path, f.offset, f.size, f.md5, this, p_replace_files, f.encrypted);
```

Comment:
```suggestion
	for (Pair<String, PackedData::PackedFile> &pair : file.get_files()) {
		PackedData::get_singleton()->add_path(p_path, pair.first, pair.second.offset, pair.second.size, pair.second.md5, this, p_replace_files, pair.second.encrypted);
```
No use of `auto` is permitted, see [here](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#auto-keyword)